### PR TITLE
Feature/add totp challenge

### PR DESCRIPTION
**Pull request by gjdodd**
#55 
#56 

*Description of changes:*
Added support for TOTP challenges, supports the existing way by defaulting to SMS, but also has an additional override method to allow setting the challenge type.

I have also done the changes to the identity provider on the other repo to automatically support the 2fa differences, currently, these changes are on my fork until this is merged as the other branch relies on changes from this branch for the RespondToMfaRquest class

I have seen other pull requests which are approved but not merged from what I can see, but these do not seem to be backwardly compatible, and also there is no other pull request on the identity provider to also update that to support the required changes.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs [hunk 1]
```diff
@@ -111,19 +111,35 @@ public class InitiateRefreshTokenAuthRequest
...
+    /// Class containing the necessary properities to respond to an MFA authentication challenge
     /// </summary>
-    public class RespondToSmsMfaRequest
+    public class RespondToMfaRequest
     {
         /// <summary>
         /// The session ID for the current authentication flow.
         /// </summary>
-        public string SessionID { get; set; }
+        public virtual string SessionID { get; set; }
```
Comment (normj) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:121]: Why was SessionID changed to be virtual?

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs [hunk 2]
```diff
@@ -111,19 +111,35 @@ public class InitiateRefreshTokenAuthRequest
...
 
         /// <summary>
         /// The MFA verification code needed to authenticate the user.
         /// </summary>
-        public string MfaCode { get; set; }
+        public virtual string MfaCode { get; set; }
+
+        /// <summary>
+        /// The MFA verification code needed to authenticate the user.
```
Comment (normj) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:129]: The description is incorrect

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs [hunk 3]
```diff
@@ -111,19 +111,35 @@ public class InitiateRefreshTokenAuthRequest
...
+        public virtual ChallengeNameType ChallengeNameType { get; set; }        
+    }
+
+    /// <summary>
+    /// Class containing the necessary properities to respond to an MFA authentication challenge
+    /// </summary>
+    public class RespondToSmsMfaRequest : RespondToMfaRequest
+    {
+        /// <summary>
+        /// The MFA verification code needed to authenticate the user.
```
Comment (normj) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:140]: The description is incorrect.

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs [hunk 4]
```diff
@@ -111,19 +111,35 @@ public class InitiateRefreshTokenAuthRequest
...
+        /// </summary>
+        public virtual ChallengeNameType ChallengeNameType { get; set; }        
+    }
+
+    /// <summary>
+    /// Class containing the necessary properities to respond to an MFA authentication challenge
+    /// </summary>
+    public class RespondToSmsMfaRequest : RespondToMfaRequest
```
Comment (normj) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:137]: I suspect we should probably just deprecate this class and encourage users to use the newer `RespondToMfaRequest`. That can be done later though.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 1 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:121]: This is a new class, and RespondToSmsMfaRequest inherits from it. This was added as a virtual so inheriting classes could override the properties or methods on this new base class

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 2 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:129]: I will get the description to match the property, probably a copy and paste error on my part

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 3 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:140]: I will get the description to match the property, probably a copy and paste error on my part

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 4 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:137]: I also spotted that the override of the property was not set correctly, I will also get that fixed and test the SMS flow again

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 3 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs]: The description has been updated in the latest commit

File: src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs (hunk 4 above)
Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:137]: I have now fixed the override of the ChallengeNameType field issue as well on the RespondToSmsMfaRequest

Comment (gjdodd) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoAuthenticationClasses.cs:137]: I have left the existing class in place for backward compatibility

**Review by normj — APPROVED** [review]
Looks good and thanks for the PR.

**Comment by normj** [comment]
This is PR is released as part of version 2.1.0. I'll work on getting your corresponding PR https://github.com/aws/aws-aspnet-cognito-identity-provider/pull/173 out soon.

**Pull request merged**